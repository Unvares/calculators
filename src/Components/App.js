import React from 'react';
import { Slider } from './Slider';
import calculators from '../Assets/calculators.json';

const App = () => <Slider calculators={calculators} />;

export { App };
