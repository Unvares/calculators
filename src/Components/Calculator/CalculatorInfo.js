import React from 'react';
import cn from 'classnames';

const CalculatorInfo = ({ formula = '', lastResponse = {} }) => {
  const renderResult = ({ value, status }) => {
    const resultClasses = cn(
      'calculator__text calculator__text_input-aligned',
      {
        calculator__text_processing: status === 'processing',
        calculator__text_error: status === 'error',
      }
    );
    return (
      <span className={resultClasses}>{value && value.toUpperCase()}</span>
    );
  };

  return (
    <section className='calculator__info'>
      <div className='calculator__line'>
        <span className='calculator__text calculator__text_engraved'>
          FORMULA
        </span>
        <span className='calculator__text calculator__text_engraved calculator__text_input-aligned'>
          {formula.toUpperCase()}
        </span>
      </div>
      <div className='calculator__line'>
        <span className='calculator__text calculator__text_engraved'>
          RESULT
        </span>
        {renderResult(lastResponse)}
      </div>
    </section>
  );
};

export { CalculatorInfo };
