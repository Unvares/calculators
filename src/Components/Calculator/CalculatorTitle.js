import React from 'react';

const CalculatorTitle = ({ children }) => (
  <h2 className='calculator__text calculator__text_engraved'>{children}</h2>
);

export { CalculatorTitle };
