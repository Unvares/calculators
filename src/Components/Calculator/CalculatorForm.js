import React from 'react';
import { Server } from '../App';

const CalculatorForm = ({ inputs, activeId, setLastResponses }) => {
  const setLastResponse = (newLastResponse) =>
    setLastResponses((state) => {
      const newState = [...state];
      newState[activeId] = newLastResponse;
      return newState;
    });
  const normalizeInput = (input) =>
    input.map((value) => (value === '' ? null : Number(value)));
  const getValues = (form) => {
    const data = new FormData(form);
    return normalizeInput([...data.values()]);
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    setLastResponse({ status: 'processing', value: 'processing...' });

    const values = getValues(event.target);
    const request = {
      method: 'POST',
      path: '',
      body: { values, id: activeId },
    };

    Server.fetch(JSON.stringify(request))
      .then((response) =>
        setLastResponse({
          value: response,
          status: 'fulfilled',
        })
      )
      .catch((message) => {
        setLastResponse({ status: 'error', value: 'error' });
        throw new Error(message);
      });
  };

  return (
    <form onSubmit={handleSubmit}>
      {inputs.map((input) => (
        <div className='calculator__line calculator__line_input' key={input}>
          <label
            className='calculator__text calculator__text_engraved'
            htmlFor={input}
          >
            {input.toUpperCase()}
          </label>
          <input
            className='calculator__input calculator__text'
            type='text'
            name={input}
            id={input}
            autoComplete='off'
          />
        </div>
      ))}
      <button type='submit' style={{ display: 'none' }}></button>
    </form>
  );
};

export { CalculatorForm };
