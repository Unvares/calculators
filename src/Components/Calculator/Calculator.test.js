import React from 'react';
import { fireEvent, render, act } from '@testing-library/react';
import { CalculatorTitle } from './CalculatorTitle';
import { CalculatorForm } from './CalculatorForm';
import { CalculatorInfo } from './CalculatorInfo';
import { Calculator } from './';
import { Server } from '../App';

jest.mock('../App');

describe('CalculatorTitle', () => {
  it('renders "Triangle Surface" title', () => {
    const {
      container: { firstChild: tree },
    } = render(<CalculatorTitle>Triangle Surface</CalculatorTitle>);
    expect(tree).toMatchSnapshot();
  });
});

describe('CalculatorForm', () => {
  const setup = () => {
    const inputs = ['height', 'base'];
    const {
      container: { firstChild: tree },
      getByLabelText,
    } = render(
      <CalculatorForm inputs={inputs} activeId={null} setLastResponses={null} />
    );

    return { tree, getByLabelText };
  };

  it('renders height and base input fields', () => {
    const { tree } = setup();
    expect(tree).toMatchSnapshot();
  });

  it('types 5 and 10 into height and base input fields', () => {
    const { tree, getByLabelText } = setup();
    const inputHeight = getByLabelText('HEIGHT');
    const inputBase = getByLabelText('BASE');

    fireEvent.change(inputHeight, { target: { value: 5 } });
    fireEvent.change(inputBase, { target: { value: 10 } });

    expect(inputHeight.value).toEqual('5');
    expect(inputBase.value).toEqual('10');
    expect(tree).toMatchSnapshot();
  });
});

describe('CalculatorInfo', () => {
  it('renders (H * B) / 2 formula', () => {
    const formula = '(H * B) / 2';
    const {
      container: { firstChild: tree },
    } = render(<CalculatorInfo formula={formula} />);

    expect(tree).toMatchSnapshot();
  });

  it('renders with processing status', () => {
    const {
      container: { firstChild: tree },
    } = render(
      <CalculatorInfo
        lastResponse={{ status: 'processing', value: 'processing...' }}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders with error status', () => {
    const {
      container: { firstChild: tree },
    } = render(
      <CalculatorInfo lastResponse={{ status: 'error', value: 'error' }} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders with fulfilled status', () => {
    const {
      container: { firstChild: tree },
    } = render(
      <CalculatorInfo lastResponse={{ status: 'fulfilled', value: '25' }} />
    );
    expect(tree).toMatchSnapshot();
  });
});

describe('Calculator', () => {
  const calculator = {
    title: 'Triangle Surface',
    inputs: ['height', 'base'],
    formula: '(H * B) / 2',
  };

  it('renders "Triangle Surface" with height and base inputs and "(H * B) / 2" formula', () => {
    const {
      container: { firstChild: tree },
    } = render(<Calculator activeId={0} calculator={calculator} />);

    expect(tree).toMatchSnapshot();
  });

  it('returns 25 on form submit with 5 and 10 as input values', async () => {
    Server.fetch = jest.fn(() => Promise.resolve('25'));
    const {
      container: { firstChild: tree },
      getByLabelText,
      findByText,
    } = render(<Calculator activeId={0} calculator={calculator} />);
    const inputHeight = getByLabelText('HEIGHT');
    const inputBase = getByLabelText('BASE');
    const submitButton = tree.querySelector('button[type="submit"]');

    fireEvent.change(inputHeight, { target: { value: 5 } });
    fireEvent.change(inputBase, { target: { value: 10 } });

    fireEvent.click(submitButton);

    expect(tree).toMatchSnapshot();

    await findByText('25');

    expect(tree).toMatchSnapshot();
  });
});
