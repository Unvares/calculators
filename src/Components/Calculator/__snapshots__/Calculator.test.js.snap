// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders "Triangle Surface" with height and base inputs and "(H * B) / 2" formula 1`] = `
<div
  class="calculator"
>
  <h2
    class="calculator__text calculator__text_engraved"
  >
    TRIANGLE SURFACE
  </h2>
  <form>
    <div
      class="calculator__line calculator__line_input"
    >
      <label
        class="calculator__text calculator__text_engraved"
        for="height"
      >
        HEIGHT
      </label>
      <input
        autocomplete="off"
        class="calculator__input calculator__text"
        id="height"
        name="height"
        type="text"
        value=""
      />
    </div>
    <div
      class="calculator__line calculator__line_input"
    >
      <label
        class="calculator__text calculator__text_engraved"
        for="base"
      >
        BASE
      </label>
      <input
        autocomplete="off"
        class="calculator__input calculator__text"
        id="base"
        name="base"
        type="text"
        value=""
      />
    </div>
    <button
      style="display: none;"
      type="submit"
    />
  </form>
  <section
    class="calculator__info"
  >
    <div
      class="calculator__line"
    >
      <span
        class="calculator__text calculator__text_engraved"
      >
        FORMULA
      </span>
      <span
        class="calculator__text calculator__text_engraved calculator__text_input-aligned"
      >
        (H * B) / 2
      </span>
    </div>
    <div
      class="calculator__line"
    >
      <span
        class="calculator__text calculator__text_engraved"
      >
        RESULT
      </span>
      <span
        class="calculator__text calculator__text_input-aligned"
      />
    </div>
  </section>
</div>
`;

exports[`Calculator returns 25 on form submit with 5 and 10 as input values 1`] = `
<div
  class="calculator"
>
  <h2
    class="calculator__text calculator__text_engraved"
  >
    TRIANGLE SURFACE
  </h2>
  <form>
    <div
      class="calculator__line calculator__line_input"
    >
      <label
        class="calculator__text calculator__text_engraved"
        for="height"
      >
        HEIGHT
      </label>
      <input
        autocomplete="off"
        class="calculator__input calculator__text"
        id="height"
        name="height"
        type="text"
        value="5"
      />
    </div>
    <div
      class="calculator__line calculator__line_input"
    >
      <label
        class="calculator__text calculator__text_engraved"
        for="base"
      >
        BASE
      </label>
      <input
        autocomplete="off"
        class="calculator__input calculator__text"
        id="base"
        name="base"
        type="text"
        value="10"
      />
    </div>
    <button
      style="display: none;"
      type="submit"
    />
  </form>
  <section
    class="calculator__info"
  >
    <div
      class="calculator__line"
    >
      <span
        class="calculator__text calculator__text_engraved"
      >
        FORMULA
      </span>
      <span
        class="calculator__text calculator__text_engraved calculator__text_input-aligned"
      >
        (H * B) / 2
      </span>
    </div>
    <div
      class="calculator__line"
    >
      <span
        class="calculator__text calculator__text_engraved"
      >
        RESULT
      </span>
      <span
        class="calculator__text calculator__text_input-aligned calculator__text_processing"
      >
        PROCESSING...
      </span>
    </div>
  </section>
</div>
`;

exports[`Calculator returns 25 on form submit with 5 and 10 as input values 2`] = `
<div
  class="calculator"
>
  <h2
    class="calculator__text calculator__text_engraved"
  >
    TRIANGLE SURFACE
  </h2>
  <form>
    <div
      class="calculator__line calculator__line_input"
    >
      <label
        class="calculator__text calculator__text_engraved"
        for="height"
      >
        HEIGHT
      </label>
      <input
        autocomplete="off"
        class="calculator__input calculator__text"
        id="height"
        name="height"
        type="text"
        value="5"
      />
    </div>
    <div
      class="calculator__line calculator__line_input"
    >
      <label
        class="calculator__text calculator__text_engraved"
        for="base"
      >
        BASE
      </label>
      <input
        autocomplete="off"
        class="calculator__input calculator__text"
        id="base"
        name="base"
        type="text"
        value="10"
      />
    </div>
    <button
      style="display: none;"
      type="submit"
    />
  </form>
  <section
    class="calculator__info"
  >
    <div
      class="calculator__line"
    >
      <span
        class="calculator__text calculator__text_engraved"
      >
        FORMULA
      </span>
      <span
        class="calculator__text calculator__text_engraved calculator__text_input-aligned"
      >
        (H * B) / 2
      </span>
    </div>
    <div
      class="calculator__line"
    >
      <span
        class="calculator__text calculator__text_engraved"
      >
        RESULT
      </span>
      <span
        class="calculator__text calculator__text_input-aligned"
      >
        25
      </span>
    </div>
  </section>
</div>
`;

exports[`CalculatorForm renders height and base input fields 1`] = `
<form>
  <div
    class="calculator__line calculator__line_input"
  >
    <label
      class="calculator__text calculator__text_engraved"
      for="height"
    >
      HEIGHT
    </label>
    <input
      autocomplete="off"
      class="calculator__input calculator__text"
      id="height"
      name="height"
      type="text"
      value=""
    />
  </div>
  <div
    class="calculator__line calculator__line_input"
  >
    <label
      class="calculator__text calculator__text_engraved"
      for="base"
    >
      BASE
    </label>
    <input
      autocomplete="off"
      class="calculator__input calculator__text"
      id="base"
      name="base"
      type="text"
      value=""
    />
  </div>
  <button
    style="display: none;"
    type="submit"
  />
</form>
`;

exports[`CalculatorForm types 5 and 10 into height and base input fields 1`] = `
<form>
  <div
    class="calculator__line calculator__line_input"
  >
    <label
      class="calculator__text calculator__text_engraved"
      for="height"
    >
      HEIGHT
    </label>
    <input
      autocomplete="off"
      class="calculator__input calculator__text"
      id="height"
      name="height"
      type="text"
      value="5"
    />
  </div>
  <div
    class="calculator__line calculator__line_input"
  >
    <label
      class="calculator__text calculator__text_engraved"
      for="base"
    >
      BASE
    </label>
    <input
      autocomplete="off"
      class="calculator__input calculator__text"
      id="base"
      name="base"
      type="text"
      value="10"
    />
  </div>
  <button
    style="display: none;"
    type="submit"
  />
</form>
`;

exports[`CalculatorInfo renders (H * B) / 2 formula 1`] = `
<section
  class="calculator__info"
>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      FORMULA
    </span>
    <span
      class="calculator__text calculator__text_engraved calculator__text_input-aligned"
    >
      (H * B) / 2
    </span>
  </div>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      RESULT
    </span>
    <span
      class="calculator__text calculator__text_input-aligned"
    />
  </div>
</section>
`;

exports[`CalculatorInfo renders with error status 1`] = `
<section
  class="calculator__info"
>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      FORMULA
    </span>
    <span
      class="calculator__text calculator__text_engraved calculator__text_input-aligned"
    />
  </div>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      RESULT
    </span>
    <span
      class="calculator__text calculator__text_input-aligned calculator__text_error"
    >
      ERROR
    </span>
  </div>
</section>
`;

exports[`CalculatorInfo renders with fulfilled status 1`] = `
<section
  class="calculator__info"
>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      FORMULA
    </span>
    <span
      class="calculator__text calculator__text_engraved calculator__text_input-aligned"
    />
  </div>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      RESULT
    </span>
    <span
      class="calculator__text calculator__text_input-aligned"
    >
      25
    </span>
  </div>
</section>
`;

exports[`CalculatorInfo renders with processing status 1`] = `
<section
  class="calculator__info"
>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      FORMULA
    </span>
    <span
      class="calculator__text calculator__text_engraved calculator__text_input-aligned"
    />
  </div>
  <div
    class="calculator__line"
  >
    <span
      class="calculator__text calculator__text_engraved"
    >
      RESULT
    </span>
    <span
      class="calculator__text calculator__text_input-aligned calculator__text_processing"
    >
      PROCESSING...
    </span>
  </div>
</section>
`;

exports[`CalculatorTitle renders "Triangle Surface" title 1`] = `
<h2
  class="calculator__text calculator__text_engraved"
>
  Triangle Surface
</h2>
`;
